src/WinRT.Runtime2/InteropServices/ExceptionHelpers.cs

```diff
@@ -0,0 +1,165 @@
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                WindowsRuntimeObjectMarshaller.Free(currentLanguageExceptionErrorInfo2Ptr);
+                _ = Marshal.Release(languageErrorInfo2Ptr);
+            }
+        }
+
+        return null;
+    }
+
+    private static unsafe Exception? GetLanguageExceptionInternal(void* languageErrorInfoPtr, int hr)
+    {
+        if (languageErrorInfoPtr == null)
+        {
+            return null;
+        }
+        void* languageExceptionPtr = null;
+        try
+        {
+            if (WellKnownErrorCodes.Succeeded(ILanguageExceptionErrorInfoVftbl.GetLanguageExceptionUnsafe(languageErrorInfoPtr, &languageExceptionPtr)))
+            {
+                if (WindowsRuntimeMarshal.TryGetManagedObject(languageExceptionPtr, out object? exception))
+                {
+                    if (RestrictedErrorInfo.GetHRForException((Exception)exception) == hr)
+                    {
+                        return (Exception)exception;
+                    }
+                }
+            }
+        }
+        finally
+        {
+            if (languageExceptionPtr != null)
+            {
+                WindowsRuntimeObjectMarshaller.Free(languageExceptionPtr);
+            }
+        }
+        return null;
+    }
+
+    internal static void AddExceptionDataForRestrictedErrorInfo(
+        this Exception ex,
+        WindowsRuntimeObjectReference restrictedErrorObject,
+        bool hasRestrictedLanguageErrorObject)
+    {
+        IDictionary dict = ex.Data;
+        if (dict != null)
+        {
+            // Keep the error object alive so that user could retrieve error information
+            // using Data["RestrictedErrorReference"]
+            dict["__RestrictedErrorObjectReference"] = restrictedErrorObject;
+            dict["__HasRestrictedLanguageErrorObject"] = hasRestrictedLanguageErrorObject;
+        }
+    }
+}
+
+internal static class ExceptionExtensions
+{
+    public static bool TryGetRestrictedLanguageErrorInfo(
+        this Exception ex,
+        out WindowsRuntimeObjectReference? restrictedErrorObject,
+        out bool isLanguageException)
+    {
+        restrictedErrorObject = null;
+        isLanguageException = false;
+
+        IDictionary dict = ex.Data;
+        if (dict != null)
+        {
+            if (dict.Contains("__RestrictedErrorObjectReference"))
+            {
+                restrictedErrorObject = dict["__RestrictedErrorObjectReference"] as WindowsRuntimeObjectReference;
+            }
+
+            if (dict.Contains("__HasRestrictedLanguageErrorObject"))
+            {
+                isLanguageException = (bool)dict["__HasRestrictedLanguageErrorObject"]!;
+            }
+
+            return restrictedErrorObject is not null;
+        }
+
+        return false;
+    }
+
+    public static void SetHResult(this Exception ex, int value)
```

Comment:
Delete this extension and just always set `HResult`, this method was only for backward-compatibility